lib/dns/server.js

```diff
@@ -383,6 +396,119 @@ class RootServer extends DNSServer {
 
     return res;
   }
+  
+  async handle(msg, rinfo) {
+    let req = null;
+    let res = null;
+
+    try {
+      req = Message.decode(msg);
+    } catch (e) {
+      this.emit('error', e);
+
+      if (msg.length < 2)
+        return;
+
+      res = new Message();
+      res.id = bio.readU16BE(msg, 0);
+      res.ra = this.ra;
+      res.qr = true;
+      res.code = codes.FORMERR;
+      this.send(null, res, rinfo);
+
+      return;
+    }
+
+    try {
+      const name = req.question[0].name;
+      const type = req.question[0].type;
+      let changed;
+      // If domain is in the form *.hsd.zone., query only the hsd.zone. subdomain
+      if (!!this.suffix) {
+        const zonePattern = `.*\.${this.suffix.split('.').join('\.')}`
+        const found = name.match(zonePattern);
+        if (found) {
+          const handshakeName = name.slice(0, -1*this.suffix.length);
+          if (this.suffix.length === name.length) {
+            handshakeName = '.';
+          }
+          req.question[0].name = handshakeName;
+          changed = true;
+        }
+      }
+
+      res = await this.answer(req, rinfo);
+
+      // Change query back to original name
+      if (changed) {
+        req.question[0].name = name;
+        res.question[0].name = name;
+        if (res.answer.length !== 0) {
+         res.answer[0].name = name;
+        }
+      }
+
+    } catch (e) {
+      this.emit('error', e);
+
+      res = new Message();
+      res.code = codes.SERVFAIL;
+
+      if (e.type === 'DNSError')
+        res.code = e.errno;
+      this.finalize(req, res);
+    }
+
+    if (res) {
+      this.emit('query', req, res, rinfo);
+      this.send(req, res, rinfo);
+      return;
+    }
+
+    res = new Response(this, req, rinfo);
+    res.setReply(req);
+    res.ra = this.ra;
+    this.emit('query', req, res, rinfo);
+  }
+  
+  send(req, res, rinfo) {
+    const {port, address, tcp} = rinfo;
+
```

Comment:
nit: general comment that primarily applies to our code and not hsd code -- spell out full words in almost all cases (`let message`, `let messageWithSignature`, etc)

To be clear -- no change required since this more closely follows handshake-org/hsd style, but just sharing.